---
title: "SGE Aging Project"
output:
  html_document:
    code_folding: hide
    toc: yes
    toc_float: yes
  pdf_document: default
---

```{r setup, include=FALSE, message=FALSE}
#cache = T, cache.lazy = FALSE,
knitr::opts_chunk$set(echo = TRUE, message = FALSE)
library(ggplot2)
library(limma)
library(edgeR)
library(EMMREML)
library(data.table)
library(cobs)
library(scales)
library(GGally)
library(ggpmisc)
library(GGally)
library(cowplot)
library(gridExtra)
library(reshape2)
library(Seurat)
library(dplyr)
library(gtools)
library(viridis)
library(SingleCellExperiment)
library(scater)
library(scran)
library(pbapply)
source('/Users/noah/Desktop/AS.helper.functions.R')
```
<!-- read in single-cell data from aging cohort and process into seurat object -->
```{r readFiles}
AS <- readRDS(file = '/Users/noah/Desktop/SGE_data_objects/AS_7SEP20.rds')
```

<!-- ```{r ASplots, warning=FALSE, message=FALSE} -->
<!-- #DimPlot(AS, reduction = "umap", label = T) -->
<!-- #DotPlot(AS, features=c("CD79A","MS4A1","NKG7","GNLY","CST3","CD14","LYZ","CD3D","IL7R","S100A4","CD8A","PPBP")) + -->
<!--   #theme(axis.text.x = element_text(angle = 90, hjust = 1)) -->
<!-- #pbmc.markers <- FindAllMarkers(AS, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25) -->
<!-- #pbmc.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_logFC) -->
<!-- ``` -->
## Age series cell clustering
Here we are matching the cell-type clustering of the SGE data as follows:

![](/Users/noah/Desktop/AS_DimPlot_celltypes.png.001.png)
<br>
<br>

## Single-cell proportions vs flow proportions
```{r, warning=FALSE, message=FALSE}
# get cell type proportions
ind_list <- unique(substr(AS@assays$RNA@counts@Dimnames[[2]],1,5))
ind_list <- gsub("_","",ind_list)
cellcount.mat <- as.data.frame(matrix(nrow = 14, ncol = 0))

for (i in 0:11){
  tmp <- SubsetData(object = AS, ident.use = i)
  cellcount_tmp <- matrix(nrow = 0,ncol = 1)
  for (ind in ind_list){
    tmp2 <- (sum(tmp$ID == ind)+1)
    cellcount_tmp <- rbind(cellcount_tmp,tmp2)
  }
  cellcount.mat <- cbind(cellcount.mat, cellcount_tmp)
}


colnames(cellcount.mat) <- c('cluster 0', 
                             'cluster 1', 
                             'cluster 2',
                             'cluster 3',
                             'cluster 4', 
                             'cluster 5',
                             'cluster 6', 
                             'cluster 7', 
                             'cluster 8',
                             'cluster 9',
                             'cluster 10',
                             'cluster 11')
rownames(cellcount.mat) <- ind_list

# combine clusters to get cell types
cellcount.mat2 <- as.data.frame(matrix(nrow = 14, ncol = 6))
#add cluster together to get cell type-level counts
cellcount.mat2[,1] <- cellcount.mat[,1] + cellcount.mat[,4]
cellcount.mat2[,2] <- cellcount.mat[,2] + cellcount.mat[,3]
cellcount.mat2[,3] <- cellcount.mat[,5] + cellcount.mat[,9]
cellcount.mat2[,4] <- cellcount.mat[,6]
cellcount.mat2[,5] <- cellcount.mat[,7]
cellcount.mat2[,6] <- cellcount.mat[,8]
colnames(cellcount.mat2) <- c('CD4', 
                              'CD8', 
                              'Bcells',
                              'NK',
                              'CD14hi', 
                              'CD14low')
rownames(cellcount.mat2) <- ind_list
```

raw counts matrix for cell types
```{r}
knitr::kable(cellcount.mat2, format = 'simple')
```
```{r}

# calculate proportions

cellprops.mat <- cellcount.mat2
for (i in 1:14){
  cellprops.mat[i,]  <- cellcount.mat2[i,] / sum(cellcount.mat2)
}

# divide by rowSums so proportions sum to one per individual
for (i in 1:14){
  cellprops.mat[i,] <- cellprops.mat[i,] / sum(cellprops.mat[i,])
}
```

cell proportions
```{r}
knitr::kable(cellprops.mat, format = "simple", digits = 4)
```
```{r}
print('rowSums: ')
print(rowSums(cellprops.mat))
```
```{r}
#read in flow proportions
flow_data <- read.delim('/Users/noah/Desktop/flow_data_raw.txt', header = TRUE)

# combine monocytes and b cells
flow_props.tmp <- flow_data[,2:11] / 100
rownames(flow_props.tmp) <- rownames(cellprops.mat)
flow_props.tmp <- flow_props.tmp[c(1:9,11:14),]
flow_props <- as.data.frame(matrix(nrow = 13, ncol = 7))
rownames(flow_props) <- rownames(flow_props.tmp)
flow_props$V1 <- flow_props.tmp$CD4.cells
flow_props$V2 <- flow_props.tmp$CD8.cells
flow_props$V3 <- flow_props.tmp$B.cells..CD8.. + flow_props.tmp$B.cells..CD8...1
flow_props$V4 <- flow_props.tmp$NK.cells
flow_props$V5 <- flow_props.tmp$Class..Mono..CD14.CD16..
flow_props$V6 <- flow_props.tmp$Act..Mono..CD14.CD16.. + flow_props.tmp$Act..Mono..CD14.CD16...1
colnames(flow_props) <- colnames(cellcount.mat2)

# subset sc proportions to match 13 individuals in flow data
cellprops.mat_sub <- subset(cellprops.mat, rownames(cellprops.mat) %in% rownames(flow_props)) 
```

Plots for SC vs. flow proportions. Here I've combined the two CD8 subclusters from SC data, and I'm using classical monocytes from flow as analog to CD14high in SC, and the combined activated monocytes in flow as analog to CD14low in SC. 
```{r}
for (i in 1:6){
  print(ggplot(flow_props, aes(x=cellprops.mat_sub[,i], y=flow_props[,i])) +
          geom_point() +
          stat_smooth(method = 'lm', color = 'black') +
          stat_fit_glance(method = "lm",
                    method.args = list(formula = y ~ x),
                    mapping = aes(label = sprintf('r^2~"="~%.3f~~~italic(P)~"="~%.2g',
                    stat(r.squared), stat(p.value))),
                    parse = TRUE) +
          xlab(paste(colnames(flow_props)[i], ' single-cell')) +
          ylab(paste(colnames(flow_props)[i], ' flow')) +
          ggtitle(colnames(flow_props)[i])
        )
}
```

Correlations between single cell and flow proportions Overall look pretty good, particularly for B, NK,and monocytes


## PCA on cell type composition
Here I'm performing a PCA analysis looking at the correlation between age and the top PCs of cell type composition. Here I'll lay out the exact steps taken, starting from an 14 x 7 matrix of cell counts with individuals on the rows and cell types on the columns:\
1. divide each cell by the sum of the entire matrix to get proportions \
2. divide each value by it's rowSum so each individuals proportions sum to one \
3. transpose matrix, logit transform proportions, and regress out group effects \
4. perform PCA (prcomp) on covariance matrix of cell proportion residuals (14 x 14) \
5. calculate correlation between age and each PC \
6. re-run PCA on rectangular matrix of cell proportion residuals to extract loadings of each cell type on PCs

```{r, message=FALSE, warning=FALSE}
# define age, group, and individuals names
age <- c(21.54,10.99,10.96,4.47,5.31,22.36,21.53,10.62,11.41,23.3,19.52,11.23,3.52,5.18)
group <- as.factor(c('1','1','1','1','1','2','2','2','2','3','3','3','3','3'))
ind_list <- unique(substr(AS@assays$RNA@counts@Dimnames[[2]],1,5))
ind_list <- gsub("_","",ind_list)

# define a list of clusters for each cell type as follows:
#CD4+ helper T: 0,3
#CD8_A+ helper T: 1
#CD8_B+ helper T: 2
#Bcells: 4,8
#NK: 5
#CD14hi: 6
#CD14low: 7

#redo cellprops to split the CD8 clusters
cellcount.mat3 <- as.data.frame(matrix(nrow = 14, ncol = 7))
#add cluster together to get cell type-level counts
cellcount.mat3[,1] <- cellcount.mat[,1] + cellcount.mat[,4]
cellcount.mat3[,2] <- cellcount.mat[,2]
cellcount.mat3[,3] <- cellcount.mat[,3]
cellcount.mat3[,4] <- cellcount.mat[,5] + cellcount.mat[,9]
cellcount.mat3[,5] <- cellcount.mat[,6]
cellcount.mat3[,6] <- cellcount.mat[,7]
cellcount.mat3[,7] <- cellcount.mat[,8]
colnames(cellcount.mat3) <- c('CD4', 
                              'CD8_A', 
                              'CD8_B', 
                              'Bcells',
                              'NK',
                              'CD14hi', 
                              'CD14low')
rownames(cellcount.mat3) <- ind_list

# calculate proportions
cellprops.mat <- cellcount.mat3
for (i in 1:14){
  cellprops.mat[i,]  <- cellcount.mat3[i,] / sum(cellcount.mat3)
}

# divide by rowSums so proportions sum to one per individual
for (i in 1:14){
  cellprops.mat[i,] <- cellprops.mat[i,] / sum(cellprops.mat[i,])
}

cellprops.mat <- t(cellprops.mat)
design <- model.matrix(~0+group)
#
v <- logit(cellprops.mat)
#v <- cellprops.mat
fit <- lmFit(v,design)
fit <- eBayes(fit)
cellprop.resid <-residuals.MArrayLM(object=fit, v)
```
```{r, message=FALSE}
pca <- prcomp(cov(cellprop.resid))
pca_age_cors <- vector()
for (i in 1:7){
  #print(cor.test(pca$x[,i],age)$p.value)
  pca_age_cors[i] <- cor.test(pca$x[,i],age)$p.value
}

ggplot() +
  geom_point(aes(x = as.factor(1:7), y = -log10(pca_age_cors), size = 1)) +
  geom_hline(yintercept = -log10(0.05), color = 'red', linetype = 'dashed', size = 1) +
  xlab('PC') +
  ylab('-log10(p)') + 
  theme(legend.position = 'none')

pca_age_cors2 <- vector()
for (i in 1:7){
  #print(cor.test(pca$x[,i],age)$p.value)
  pca_age_cors2[i] <- summary(pca)$importance[2,i]
}

ggplot() +
  geom_point(aes(x = as.factor(1:7), y = pca_age_cors2)) +
  geom_line(aes(x = as.factor(1:7), y = pca_age_cors2, group = 1)) +
  xlab('PC') +
  ylab('PVE by PC')

qplot(pca$x[,1],pca$x[,2],col=age,size=1.5)+
  xlab(paste("PC1","\nPVE =",round(summary(pca)$imp[2,1],2)))+
  ylab(paste("PC2","\nPVE =",round(summary(pca)$imp[2,2],2))) 

#qplot(age, pca$x[,1], col=age,size=1) + xlab('age') + ylab(paste("PC1","\nPVE =",round(summary(pca)$imp[2,1],2))) + stat_smooth(method = 'lm', #color='black', size = 1) 
qplot(age, pca$x[,3], col=age,size=1) + xlab('age') + ylab(paste("PC3","\nPVE =",round(summary(pca)$imp[2,3],2))) + stat_smooth(method = 'lm', color='black', size = 1) 
#+ scale_color_viridis_c(option = 'E')
#ggsave(filename = '/Users/noah/Desktop/agePC1.png', scale = 0.5)
```

### Loadings on PCs
```{r, message=FALSE, warning=FALSE}
knitr::kable(prcomp(t(cellprop.resid))$rotation, format = "simple", digits = 2)
```

Age is most strongly related to PC3, which explains ~12% of variance in cell composition. Monocytes and B cells load most heavily on PC3. 


## PCA on pseudobulk gene expression (total PBMCs)
```{r generate pseudobulk means object, cache=TRUE, cache.lazy=FALSE}
#### Now let's look at the PCA for the gene expression data itself
sce <- as.SingleCellExperiment(AS, assay="RNA")
## QC metrics ##
#names(assays(sce))
sce <- addPerCellQC(sce)
sce <- addPerFeatureQC(sce)
logCPM_filter <- 0.2
keep_gene <- rowMeans(log2(calculateCPM(sce) + 1)) > logCPM_filter
sce_filt <- sce[keep_gene,]
sce_filt <- computeSumFactors(sce_filt)
sce_filt$size_factor <- sizeFactors(sce_filt)
colData(sce_filt)$group <- as.factor(colData(sce_filt)$group)
sce_filt <- sce_filt[, (sizeFactors(sce_filt) < 8 & sizeFactors(sce_filt) > 0.125)]

sce_filt <- logNormCounts(sce_filt, pseudo_count = 1)
normalized_data <- exprs(sce_filt)
meta_data <- as.data.frame(colData(sce_filt))
meta_data$sample_condition <- paste(meta_data$ID, meta_data$trt, sep="")
sample_colname <- "sample_condition"
IDs <- as.data.frame(meta_data)[, sample_colname]
unique_ID_list <- as.list(unique(IDs))

pseudobulk <- as.data.frame(pblapply(unique_ID_list, FUN = function(x){rowMeans(normalized_data[,IDs == x, drop = FALSE])}))

colnames(pseudobulk) <- unique(IDs)
rownames(pseudobulk) <- rownames(normalized_data)
pseudobulk <- as.data.frame(pseudobulk)
object_pseudobulkMeans <- pseudobulk

group <- as.factor(c('1','1','1','1','1','2','2','2','2','3','3','3','3','3'))
design <- model.matrix(~0+group)
fit <-lmFit(object_pseudobulkMeans,design)
fit <- eBayes(fit)
residuals <-residuals.MArrayLM(object=fit, object_pseudobulkMeans)
residuals <- residuals[order(rownames(residuals)),]
residuals <- as.data.frame(residuals)
```
```{r PCA on gene expression}
pca_AS <- prcomp(cov(residuals))
#for (i in 1:10){
#  print(cor.test(pca_AS$x[,i],age)$p.value)
#}

pca_pvals <- matrix(nrow = 10, ncol = 2)
for (p in 1:10){
  pca_pvals[p,1] <- cor.test(pca_AS$x[,p],age)$p.value
}
for (p in 1:10){
  pca_pvals[p,2] <- summary(pca_AS)$importance[2,p]
}

ggplot() +
  geom_point(aes(x = as.factor(1:10), y = -log10(pca_pvals[,1]), size = 1)) +
  geom_hline(yintercept = -log10(0.05), color = 'red', linetype = 'dashed', size = 1) +
  xlab('PC') +
  ylab('-log10(p)') + 
  theme(legend.position = 'none')

ggplot() +
  geom_point(aes(x = as.factor(1:10), y = pca_pvals[,2])) +
  geom_line(aes(x = as.factor(1:10), y = pca_pvals[,2], group = 1)) +
  xlab('PC') +
  ylab('PVE by PC')

#qplot(pca_AS$x[,1],pca_AS$x[,2],col=age,size=1.5)+xlab(paste("PC1","\nPVE =",round(summary(pca_AS)$imp[2,1],2)))+ylab(paste("PC2","\nPVE =",round(summary(pca_AS)$imp[2,2],2)))
qplot(age, pca_AS$x[,3], col=age,size=1) + xlab('age') + ylab(paste("PC3","\nPVE =",round(summary(pca_AS)$imp[2,3],2))) + stat_smooth(method = 'lm', color='black', size = 1)
```

Now I'll run a linear model to test the association between age and PC3 of gene expression while controlling for the top 2 PCs of cell composition
```{r}
knitr::kable(summary(lm(pca_AS$x[,3] ~ age + pca$x[,1] + pca$x[,2]))$coefficients, format = 'simple', digits = 3)
```

It looks like controlling for cell composition does not improve the relationshipnship between age and PC1 of gene expression in total PBMC pseudobulk
<br>

## PCA on gene expression in individual cell types
```{r CD4+ T cells, warning=FALSE, message=FALSE, cache=T, cache.lazy=FALSE}
tmp_pseudobulkMeans <- generatePseudobulkMeans(clusterNum = c(0,3))
tmp_residuals <- generatePseudobulkMeansResiduals(object_pseudobulk = tmp_pseudobulkMeans)
pca_AS <- prcomp(cov(tmp_residuals))
pca_pvals <- matrix(nrow = 10, ncol = 2)
for (p in 1:10){
  pca_pvals[p,1] <- cor.test(pca_AS$x[,p],age)$p.value
}
for (p in 1:10){
  pca_pvals[p,2] <- summary(pca_AS)$importance[2,p]
}
p1 <- ggplot() +
        geom_point(aes(x = as.factor(1:10), y = -log10(pca_pvals[,1]), size = 1)) +
        geom_hline(yintercept = -log10(0.05), color = 'red', linetype = 'dashed', size = 1) +
        xlab('PC') +
        ylab('-log10(p)') + 
        theme(legend.position = 'none') +
        ggtitle(label = '', subtitle = 'CD4+ T')
p2 <- ggplot() +
        geom_point(aes(x = as.factor(1:10), y = pca_pvals[,2])) +
        geom_line(aes(x = as.factor(1:10), y = pca_pvals[,2], group = 1)) +
        xlab('PC') +
        ylab('PVE by PC') +
        ggtitle(label = '', subtitle = 'CD4+ T')
p_all <- plot_grid(p1,p2)
p_all
```
```{r CD8_A+ T cells, warning=FALSE, message=FALSE, cache=T, cache.lazy=FALSE}
tmp_pseudobulkMeans <- generatePseudobulkMeans(clusterNum = c(1))
tmp_residuals <- generatePseudobulkMeansResiduals(object_pseudobulk = tmp_pseudobulkMeans)
pca_AS <- prcomp(cov(tmp_residuals))
pca_pvals <- matrix(nrow = 10, ncol = 2)
for (p in 1:10){
  pca_pvals[p,1] <- cor.test(pca_AS$x[,p],age)$p.value
}
for (p in 1:10){
  pca_pvals[p,2] <- summary(pca_AS)$importance[2,p]
}
p1 <- ggplot() +
        geom_point(aes(x = as.factor(1:10), y = -log10(pca_pvals[,1]), size = 1)) +
        geom_hline(yintercept = -log10(0.05), color = 'red', linetype = 'dashed', size = 1) +
        xlab('PC') +
        ylab('-log10(p)') + 
        theme(legend.position = 'none') +
        ggtitle(label = '', subtitle = 'CD8_A+ T')
p2 <- ggplot() +
        geom_point(aes(x = as.factor(1:10), y = pca_pvals[,2])) +
        geom_line(aes(x = as.factor(1:10), y = pca_pvals[,2], group = 1)) +
        xlab('PC') +
        ylab('PVE by PC') +
        ggtitle(label = '', subtitle = 'CD8_A+ T')
p_all <- plot_grid(p1,p2)
p_all
```
```{r CD8_B+ T cells, warning=FALSE, message=FALSE, cache=T, cache.lazy=FALSE}
tmp_pseudobulkMeans <- generatePseudobulkMeans(clusterNum = c(2))
tmp_residuals <- generatePseudobulkMeansResiduals(object_pseudobulk = tmp_pseudobulkMeans)
pca_AS <- prcomp(cov(tmp_residuals))
pca_pvals <- matrix(nrow = 10, ncol = 2)
for (p in 1:10){
  pca_pvals[p,1] <- cor.test(pca_AS$x[,p],age)$p.value
}
for (p in 1:10){
  pca_pvals[p,2] <- summary(pca_AS)$importance[2,p]
}
p1 <- ggplot() +
        geom_point(aes(x = as.factor(1:10), y = -log10(pca_pvals[,1]), size = 1)) +
        geom_hline(yintercept = -log10(0.05), color = 'red', linetype = 'dashed', size = 1) +
        xlab('PC') +
        ylab('-log10(p)') + 
        theme(legend.position = 'none') +
        ggtitle(label = '', subtitle = 'CD8_B+ T')
p2 <- ggplot() +
        geom_point(aes(x = as.factor(1:10), y = pca_pvals[,2])) +
        geom_line(aes(x = as.factor(1:10), y = pca_pvals[,2], group = 1)) +
        xlab('PC') +
        ylab('PVE by PC') +
        ggtitle(label = '', subtitle = 'CD8_B+ T')
p_all <- plot_grid(p1,p2)
p_all
```
```{r NK, warning=FALSE, message=FALSE, cache=T, cache.lazy=FALSE}
tmp_pseudobulkMeans <- generatePseudobulkMeans(clusterNum = c(5))
tmp_residuals <- generatePseudobulkMeansResiduals(object_pseudobulk = tmp_pseudobulkMeans)
pca_AS <- prcomp(cov(tmp_residuals))
pca_pvals <- matrix(nrow = 10, ncol = 2)
for (p in 1:10){
  pca_pvals[p,1] <- cor.test(pca_AS$x[,p],age)$p.value
}
for (p in 1:10){
  pca_pvals[p,2] <- summary(pca_AS)$importance[2,p]
}
p1 <- ggplot() +
        geom_point(aes(x = as.factor(1:10), y = -log10(pca_pvals[,1]), size = 1)) +
        geom_hline(yintercept = -log10(0.05), color = 'red', linetype = 'dashed', size = 1) +
        xlab('PC') +
        ylab('-log10(p)') + 
        theme(legend.position = 'none') +
        ggtitle(label = '', subtitle = 'NK')
p2 <- ggplot() +
        geom_point(aes(x = as.factor(1:10), y = pca_pvals[,2])) +
        geom_line(aes(x = as.factor(1:10), y = pca_pvals[,2], group = 1)) +
        xlab('PC') +
        ylab('PVE by PC') +
        ggtitle(label = '', subtitle = 'NK')
p_all <- plot_grid(p1,p2)
p_all
```
```{r B cells, warning=FALSE, message=FALSE, cache=T, cache.lazy=FALSE}
tmp_pseudobulkMeans <- generatePseudobulkMeans(clusterNum = c(4,8))
tmp_residuals <- generatePseudobulkMeansResiduals(object_pseudobulk = tmp_pseudobulkMeans)
pca_AS <- prcomp(cov(tmp_residuals))
pca_pvals <- matrix(nrow = 10, ncol = 2)
for (p in 1:10){
  pca_pvals[p,1] <- cor.test(pca_AS$x[,p],age)$p.value
}
for (p in 1:10){
  pca_pvals[p,2] <- summary(pca_AS)$importance[2,p]
}
p1 <- ggplot() +
        geom_point(aes(x = as.factor(1:10), y = -log10(pca_pvals[,1]), size = 1)) +
        geom_hline(yintercept = -log10(0.05), color = 'red', linetype = 'dashed', size = 1) +
        xlab('PC') +
        ylab('-log10(p)') + 
        theme(legend.position = 'none') +
        ggtitle(label = '', subtitle = 'Bcells')
p2 <- ggplot() +
        geom_point(aes(x = as.factor(1:10), y = pca_pvals[,2])) +
        geom_line(aes(x = as.factor(1:10), y = pca_pvals[,2], group = 1)) +
        xlab('PC') +
        ylab('PVE by PC') +
        ggtitle(label = '', subtitle = 'Bcells')
p_all <- plot_grid(p1,p2)
p_all
```
```{r CD14high monocytes, warning=FALSE, message=FALSE, cache=T, cache.lazy=FALSE}
tmp_pseudobulkMeans <- generatePseudobulkMeans(clusterNum = c(6))
tmp_residuals <- generatePseudobulkMeansResiduals(object_pseudobulk = tmp_pseudobulkMeans)
pca_AS <- prcomp(cov(tmp_residuals))
pca_pvals <- matrix(nrow = 10, ncol = 2)
for (p in 1:10){
  pca_pvals[p,1] <- cor.test(pca_AS$x[,p],age)$p.value
}
for (p in 1:10){
  pca_pvals[p,2] <- summary(pca_AS)$importance[2,p]
}
p1 <- ggplot() +
        geom_point(aes(x = as.factor(1:10), y = -log10(pca_pvals[,1]), size = 1)) +
        geom_hline(yintercept = -log10(0.05), color = 'red', linetype = 'dashed', size = 1) +
        xlab('PC') +
        ylab('-log10(p)') + 
        theme(legend.position = 'none') +
        ggtitle(label = '', subtitle = 'CD14high')
p2 <- ggplot() +
        geom_point(aes(x = as.factor(1:10), y = pca_pvals[,2])) +
        geom_line(aes(x = as.factor(1:10), y = pca_pvals[,2], group = 1)) +
        xlab('PC') +
        ylab('PVE by PC') +
        ggtitle(label = '', subtitle = 'CD14high')
p_all <- plot_grid(p1,p2)
p_all
```
```{r CD14low monocytes, warning=FALSE, message=FALSE, cache=T, cache.lazy=FALSE}
tmp_pseudobulkMeans <- generatePseudobulkMeans(clusterNum = c(7))
tmp_residuals <- generatePseudobulkMeansResiduals(object_pseudobulk = tmp_pseudobulkMeans)
pca_AS <- prcomp(cov(tmp_residuals))
pca_pvals <- matrix(nrow = 10, ncol = 2)
for (p in 1:10){
  pca_pvals[p,1] <- cor.test(pca_AS$x[,p],age)$p.value
}
for (p in 1:10){
  pca_pvals[p,2] <- summary(pca_AS)$importance[2,p]
}
p1 <- ggplot() +
        geom_point(aes(x = as.factor(1:10), y = -log10(pca_pvals[,1]), size = 1)) +
        geom_hline(yintercept = -log10(0.05), color = 'red', linetype = 'dashed', size = 1) +
        xlab('PC') +
        ylab('-log10(p)') + 
        theme(legend.position = 'none') +
        ggtitle(label = '', subtitle = 'CD14low')
p2 <- ggplot() +
        geom_point(aes(x = as.factor(1:10), y = pca_pvals[,2])) +
        geom_line(aes(x = as.factor(1:10), y = pca_pvals[,2], group = 1)) +
        xlab('PC') +
        ylab('PVE by PC') +
        ggtitle(label = '', subtitle = 'CD14low')
p_all <- plot_grid(p1,p2)
p_all
```

<!-- ## Gene-level age effects models for cell types -->
<!-- ### total PBMC -->
<!-- ```{r total PBMC model, message=FALSE, warning=FALSE} -->
<!-- tmp_pseudobulk <- generatePseudobulkMeans(clusterNum = c(1:8)) -->
<!-- tmp_residuals <- generatePseudobulkMeansResiduals(tmp_pseudobulk) -->
<!-- tmp_modelRes <- modelResiduals(tmp_residuals) -->
<!-- shuffled_age_pvals <- generatePermutations(tmp_residuals,numIters = 100) -->
<!-- tmp_modelRes_fdr <- perm.fdr(tmp_modelRes,shuffled_age_pvals,'p_value_age','age') -->
<!-- p1 <- plotAgeCellCountCorrelations(clusterNum = c(1:8)) -->
<!-- p2 <- plotBetaBias(tmp_modelRes,tmp_pseudobulk) -->
<!-- colnames(shuffled_age_pvals) <- 1:100 -->
<!-- permsMelt <- reshape2::melt(shuffled_age_pvals) -->
<!-- p3 <- ggplot() +  -->
<!--   geom_histogram(aes(x = tmp_modelRes_fdr$p_value_age, y=..density..), color = 'gray30', fill = 'white', bins = 100) + -->
<!--   theme_minimal() + -->
<!--   geom_density(aes(x=permsMelt$value), color='black', size=1) + -->
<!--   xlab('pval distribution') + -->
<!--   ggtitle(label = '', subtitle = paste('# genes FDR < 0.1 = ',sum(tmp_modelRes_fdr$fdr_age < 0.1), sep = '')) -->
<!-- p4 <- ggplot(permsMelt, aes(x=value, color=variable)) +  -->
<!--   geom_density() +  -->
<!--   theme_minimal() + -->
<!--   NoLegend() + -->
<!--   xlim(-0.04,1.02) -->
<!-- p_all <- plot_grid(p1,p2,p3,p4) -->
<!--   #ggsave(paste('/Users/noah/Desktop/cluster_',i,'_Means.dashboard.png',sep=''),plot=p_all, width = 12, height = 8, dpi = 300, units = 'in') -->
<!-- p_all -->
<!-- assign(value = tmp_modelRes_fdr, x = 'totalPBMC_modelRes_fdr') -->
<!-- ``` -->

<!-- ### CD4+ T cells -->
<!-- ```{r CD4+ T model, message=FALSE, warning=FALSE} -->
<!-- tmp_pseudobulk <- generatePseudobulkMeans(clusterNum = c(0,3)) -->
<!-- tmp_residuals <- generatePseudobulkMeansResiduals(tmp_pseudobulk) -->
<!-- tmp_modelRes <- modelResiduals(tmp_residuals) -->
<!-- shuffled_age_pvals <- generatePermutations(tmp_residuals,numIters = 100) -->
<!-- tmp_modelRes_fdr <- perm.fdr(tmp_modelRes,shuffled_age_pvals,'p_value_age','age') -->
<!-- p1 <- plotAgeCellCountCorrelations(clusterNum = c(0,3)) -->
<!-- p2 <- plotBetaBias(tmp_modelRes,tmp_pseudobulk) -->
<!-- colnames(shuffled_age_pvals) <- 1:100 -->
<!-- permsMelt <- reshape2::melt(shuffled_age_pvals) -->
<!-- p3 <- ggplot() +  -->
<!--   geom_histogram(aes(x = tmp_modelRes_fdr$p_value_age, y=..density..), color = 'gray30', fill = 'white', bins = 100) + -->
<!--   theme_minimal() + -->
<!--   geom_density(aes(x=permsMelt$value), color='black', size=1) + -->
<!--   xlab('pval distribution') + -->
<!--   ggtitle(label = '', subtitle = paste('# genes FDR < 0.1 = ',sum(tmp_modelRes_fdr$fdr_age < 0.1), sep = '')) -->
<!-- p4 <- ggplot(permsMelt, aes(x=value, color=variable)) +  -->
<!--   geom_density() +  -->
<!--   theme_minimal() + -->
<!--   NoLegend() + -->
<!--   xlim(-0.04,1.02) -->
<!-- p_all <- plot_grid(p1,p2,p3,p4) -->
<!--   #ggsave(paste('/Users/noah/Desktop/cluster_',i,'_Means.dashboard.png',sep=''),plot=p_all, width = 12, height = 8, dpi = 300, units = 'in') -->
<!-- p_all -->
<!-- assign(value = tmp_modelRes_fdr, x = 'CD4_modelRes_fdr') -->
<!-- ``` -->

<!-- ### CD8_A+ T cells -->
<!-- ```{r CD8_A+ T model, message=FALSE, warning=FALSE} -->
<!-- tmp_pseudobulk <- generatePseudobulkMeans(clusterNum = c(1)) -->
<!-- tmp_residuals <- generatePseudobulkMeansResiduals(tmp_pseudobulk) -->
<!-- tmp_modelRes <- modelResiduals(tmp_residuals) -->
<!-- shuffled_age_pvals <- generatePermutations(tmp_residuals,numIters = 100) -->
<!-- tmp_modelRes_fdr <- perm.fdr(tmp_modelRes,shuffled_age_pvals,'p_value_age','age') -->
<!-- p1 <- plotAgeCellCountCorrelations(clusterNum = c(1)) -->
<!-- p2 <- plotBetaBias(tmp_modelRes,tmp_pseudobulk) -->
<!-- colnames(shuffled_age_pvals) <- 1:100 -->
<!-- permsMelt <- reshape2::melt(shuffled_age_pvals) -->
<!-- p3 <- ggplot() +  -->
<!--   geom_histogram(aes(x = tmp_modelRes_fdr$p_value_age, y=..density..), color = 'gray30', fill = 'white', bins = 100) + -->
<!--   theme_minimal() + -->
<!--   geom_density(aes(x=permsMelt$value), color='black', size=1) + -->
<!--   xlab('pval distribution')+ -->
<!--   ggtitle(label = '', subtitle = paste('# genes FDR < 0.1 = ',sum(tmp_modelRes_fdr$fdr_age < 0.1), sep = '')) -->
<!-- p4 <- ggplot(permsMelt, aes(x=value, color=variable)) +  -->
<!--   geom_density() +  -->
<!--   theme_minimal() + -->
<!--   NoLegend() + -->
<!--   xlim(-0.04,1.02) -->
<!-- p_all <- plot_grid(p1,p2,p3,p4) -->
<!--   #ggsave(paste('/Users/noah/Desktop/cluster_',i,'_Means.dashboard.png',sep=''),plot=p_all, width = 12, height = 8, dpi = 300, units = 'in') -->
<!-- p_all -->
<!-- assign(value = tmp_modelRes_fdr, x = 'CD8_modelRes_fdr') -->
<!-- ``` -->

<!-- ### CD8_B+ T cells -->
<!-- ```{r CD8_B+ T model, message=FALSE, warning=FALSE} -->
<!-- tmp_pseudobulk <- generatePseudobulkMeans(clusterNum = c(2)) -->
<!-- tmp_residuals <- generatePseudobulkMeansResiduals(tmp_pseudobulk) -->
<!-- tmp_modelRes <- modelResiduals(tmp_residuals) -->
<!-- shuffled_age_pvals <- generatePermutations(tmp_residuals,numIters = 100) -->
<!-- tmp_modelRes_fdr <- perm.fdr(tmp_modelRes,shuffled_age_pvals,'p_value_age','age') -->
<!-- p1 <- plotAgeCellCountCorrelations(clusterNum = c(2)) -->
<!-- p2 <- plotBetaBias(tmp_modelRes,tmp_pseudobulk) -->
<!-- colnames(shuffled_age_pvals) <- 1:100 -->
<!-- permsMelt <- reshape2::melt(shuffled_age_pvals) -->
<!-- p3 <- ggplot() +  -->
<!--   geom_histogram(aes(x = tmp_modelRes_fdr$p_value_age, y=..density..), color = 'gray30', fill = 'white', bins = 100) + -->
<!--   theme_minimal() + -->
<!--   geom_density(aes(x=permsMelt$value), color='black', size=1) + -->
<!--   xlab('pval distribution')+ -->
<!--   ggtitle(label = '', subtitle = paste('# genes FDR < 0.1 = ',sum(tmp_modelRes_fdr$fdr_age < 0.1), sep = '')) -->
<!-- p4 <- ggplot(permsMelt, aes(x=value, color=variable)) +  -->
<!--   geom_density() +  -->
<!--   theme_minimal() + -->
<!--   NoLegend() + -->
<!--   xlim(-0.04,1.02) -->
<!-- p_all <- plot_grid(p1,p2,p3,p4) -->
<!--   #ggsave(paste('/Users/noah/Desktop/cluster_',i,'_Means.dashboard.png',sep=''),plot=p_all, width = 12, height = 8, dpi = 300, units = 'in') -->
<!-- p_all -->
<!-- assign(value = tmp_modelRes_fdr, x = 'CD8_modelRes_fdr') -->
<!-- ``` -->

<!-- ### NK cells -->
<!-- ```{r NK model, message=FALSE, warning=FALSE} -->
<!-- tmp_pseudobulk <- generatePseudobulkMeans(clusterNum = c(5)) -->
<!-- tmp_residuals <- generatePseudobulkMeansResiduals(tmp_pseudobulk) -->
<!-- tmp_modelRes <- modelResiduals(tmp_residuals) -->
<!-- shuffled_age_pvals <- generatePermutations(tmp_residuals,numIters = 100) -->
<!-- tmp_modelRes_fdr <- perm.fdr(tmp_modelRes,shuffled_age_pvals,'p_value_age','age') -->
<!-- p1 <- plotAgeCellCountCorrelations(clusterNum = c(5)) -->
<!-- p2 <- plotBetaBias(tmp_modelRes,tmp_pseudobulk) -->
<!-- colnames(shuffled_age_pvals) <- 1:100 -->
<!-- permsMelt <- reshape2::melt(shuffled_age_pvals) -->
<!-- p3 <- ggplot() +  -->
<!--   geom_histogram(aes(x = tmp_modelRes_fdr$p_value_age, y=..density..), color = 'gray30', fill = 'white', bins = 100) + -->
<!--   theme_minimal() + -->
<!--   geom_density(aes(x=permsMelt$value), color='black', size=1) + -->
<!--   xlab('pval distribution')+ -->
<!--   ggtitle(label = '', subtitle = paste('# genes FDR < 0.1 = ',sum(tmp_modelRes_fdr$fdr_age < 0.1), sep = '')) -->
<!-- p4 <- ggplot(permsMelt, aes(x=value, color=variable)) +  -->
<!--   geom_density() +  -->
<!--   theme_minimal() + -->
<!--   NoLegend() + -->
<!--   xlim(-0.04,1.02) -->
<!-- p_all <- plot_grid(p1,p2,p3,p4) -->
<!--   #ggsave(paste('/Users/noah/Desktop/cluster_',i,'_Means.dashboard.png',sep=''),plot=p_all, width = 12, height = 8, dpi = 300, units = 'in') -->
<!-- p_all -->
<!-- assign(value = tmp_modelRes_fdr, x = 'NK_modelRes_fdr') -->
<!-- ``` -->

<!-- ### B cells -->
<!-- ```{r Bcells model, message=FALSE, warning=FALSE} -->
<!-- tmp_pseudobulk <- generatePseudobulkMeans(clusterNum = c(4,8)) -->
<!-- tmp_residuals <- generatePseudobulkMeansResiduals(tmp_pseudobulk) -->
<!-- tmp_modelRes <- modelResiduals(tmp_residuals) -->
<!-- shuffled_age_pvals <- generatePermutations(tmp_residuals,numIters = 100) -->
<!-- tmp_modelRes_fdr <- perm.fdr(tmp_modelRes,shuffled_age_pvals,'p_value_age','age') -->
<!-- p1 <- plotAgeCellCountCorrelations(clusterNum = c(4,8)) -->
<!-- p2 <- plotBetaBias(tmp_modelRes,tmp_pseudobulk) -->
<!-- colnames(shuffled_age_pvals) <- 1:100 -->
<!-- permsMelt <- reshape2::melt(shuffled_age_pvals) -->
<!-- p3 <- ggplot() +  -->
<!--   geom_histogram(aes(x = tmp_modelRes_fdr$p_value_age, y=..density..), color = 'gray30', fill = 'white', bins = 100) + -->
<!--   theme_minimal() + -->
<!--   geom_density(aes(x=permsMelt$value), color='black', size=1) + -->
<!--   xlab('pval distribution')+ -->
<!--   ggtitle(label = '', subtitle = paste('# genes FDR < 0.1 = ',sum(tmp_modelRes_fdr$fdr_age < 0.1), sep = '')) -->
<!-- p4 <- ggplot(permsMelt, aes(x=value, color=variable)) +  -->
<!--   geom_density() +  -->
<!--   theme_minimal() + -->
<!--   NoLegend() + -->
<!--   xlim(-0.04,1.02) -->
<!-- p_all <- plot_grid(p1,p2,p3,p4) -->
<!--   #ggsave(paste('/Users/noah/Desktop/cluster_',i,'_Means.dashboard.png',sep=''),plot=p_all, width = 12, height = 8, dpi = 300, units = 'in') -->
<!-- p_all -->
<!-- assign(value = tmp_modelRes_fdr, x = 'Bcells_modelRes_fdr') -->
<!-- ``` -->

<!-- ### CD14hi monocytes -->
<!-- ```{r CD14high model, message=FALSE, warning=FALSE} -->
<!-- tmp_pseudobulk <- generatePseudobulkMeans(clusterNum = c(6)) -->
<!-- tmp_residuals <- generatePseudobulkMeansResiduals(tmp_pseudobulk) -->
<!-- tmp_modelRes <- modelResiduals(tmp_residuals) -->
<!-- shuffled_age_pvals <- generatePermutations(tmp_residuals,numIters = 100) -->
<!-- tmp_modelRes_fdr <- perm.fdr(tmp_modelRes,shuffled_age_pvals,'p_value_age','age') -->
<!-- p1 <- plotAgeCellCountCorrelations(clusterNum = c(6)) -->
<!-- p2 <- plotBetaBias(tmp_modelRes,tmp_pseudobulk) -->
<!-- colnames(shuffled_age_pvals) <- 1:100 -->
<!-- permsMelt <- reshape2::melt(shuffled_age_pvals) -->
<!-- p3 <- ggplot() +  -->
<!--   geom_histogram(aes(x = tmp_modelRes_fdr$p_value_age, y=..density..), color = 'gray30', fill = 'white', bins = 100) + -->
<!--   theme_minimal() + -->
<!--   geom_density(aes(x=permsMelt$value), color='black', size=1) + -->
<!--   xlab('pval distribution')+ -->
<!--   ggtitle(label = '', subtitle = paste('# genes FDR < 0.1 = ',sum(tmp_modelRes_fdr$fdr_age < 0.1), sep = '')) -->
<!-- p4 <- ggplot(permsMelt, aes(x=value, color=variable)) +  -->
<!--   geom_density() +  -->
<!--   theme_minimal() + -->
<!--   NoLegend() + -->
<!--   xlim(-0.04,1.02) -->
<!-- p_all <- plot_grid(p1,p2,p3,p4) -->
<!--   #ggsave(paste('/Users/noah/Desktop/cluster_',i,'_Means.dashboard.png',sep=''),plot=p_all, width = 12, height = 8, dpi = 300, units = 'in') -->
<!-- p_all -->
<!-- assign(value = tmp_modelRes_fdr, x = 'CD14high_modelRes_fdr') -->
<!-- ``` -->

<!-- ### CD14low monocytes -->
<!-- ```{r CD14low model, message=FALSE, warning=FALSE} -->
<!-- tmp_pseudobulk <- generatePseudobulkMeans(clusterNum = c(7)) -->
<!-- tmp_residuals <- generatePseudobulkMeansResiduals(tmp_pseudobulk) -->
<!-- tmp_modelRes <- modelResiduals(tmp_residuals) -->
<!-- shuffled_age_pvals <- generatePermutations(tmp_residuals,numIters = 100) -->
<!-- tmp_modelRes_fdr <- perm.fdr(tmp_modelRes,shuffled_age_pvals,'p_value_age','age') -->
<!-- p1 <- plotAgeCellCountCorrelations(clusterNum = c(7)) -->
<!-- p2 <- plotBetaBias(tmp_modelRes,tmp_pseudobulk) -->
<!-- colnames(shuffled_age_pvals) <- 1:100 -->
<!-- permsMelt <- reshape2::melt(shuffled_age_pvals) -->
<!-- p3 <- ggplot() +  -->
<!--   geom_histogram(aes(x = tmp_modelRes_fdr$p_value_age, y=..density..), color = 'gray30', fill = 'white', bins = 100) + -->
<!--   theme_minimal() + -->
<!--   geom_density(aes(x=permsMelt$value), color='black', size=1) + -->
<!--   xlab('pval distribution')+ -->
<!--   ggtitle(label = '', subtitle = paste('# genes FDR < 0.1 = ',sum(tmp_modelRes_fdr$fdr_age < 0.1), sep = '')) -->
<!-- p4 <- ggplot(permsMelt, aes(x=value, color=variable)) +  -->
<!--   geom_density() +  -->
<!--   theme_minimal() + -->
<!--   NoLegend() + -->
<!--   xlim(-0.04,1.02) -->
<!-- p_all <- plot_grid(p1,p2,p3,p4) -->
<!--   #ggsave(paste('/Users/noah/Desktop/cluster_',i,'_Means.dashboard.png',sep=''),plot=p_all, width = 12, height = 8, dpi = 300, units = 'in') -->
<!-- p_all -->
<!-- assign(value = tmp_modelRes_fdr, x = 'CD14low_modelRes_fdr') -->
<!-- ``` -->

## Gene-level age effects models across total PBMCs and individual cell types

Total PBMCs
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/totalPBMC_Means.dashboard.png)

CD4 T cells
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/CD4_Means.dashboard.png)

CD8_A T cells
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/CD8_A_Means.dashboard.png)

CD8_B T cells
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/CD8_B_Means.dashboard.png)

NK cells
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/NK_Means.dashboard.png)

B cells
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/Bcell_Means.dashboard.png)

CD14high monocytes
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/CD14high_Means.dashboard.png)

CD14low monocytes
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/CD14low_Means.dashboard.png)


<br>
The overall picture from gene-level modeling of age effects suggest we are power limited to detect age effects in this data set. Note: I also confirmed this in an independent data set of bulk RNAseq from Amboseli baboons in which there are string age effects. I downsampled that data set to match the sex, sample size, and age distribution of out data set and found that age signature no longer detectable. 

To address these power limitations, we will try two approaches: 1) generate an ascertainment set of age-related genes in rhesus macaques (I'll use data on age effects in the macaques from Cayo Santiago provided to me by Kenny Chiou). Re-reun the gene-level models on this ascertainment set to see if we can detect an age signature in our data (in reality I will just subset the original model output to include the ascertainment set before calculating the FDRs). 2) Leverage information from across all of the individual cell types using a meta-analytic approach (MASH) to determine if we see a shared signature of age across cell types. 


### Gene-level modeling in ascertainment set
In the Cayo data, there are 153 significant age effect genes at an FDR < 0.1. I think we want to model more that that so I'll take genes with an FDR < 0.3, which gives us 450 genes. Note: each cell type in our single-cell data set will have a slightly different intersection gene set with the cayo gene set. 
Some metadata info re: the cayo data set: 48 individuals, both sexes, age range 4-24 yrs.

Cayo data preprocessing
```{r}
# read in Cayo data; convert ensembl IDs to gene symbols to match our data; get gene list for significant genes
cayo_data <- read.table('/Users/noah/Downloads/rna_age_effects.txt', header = T)
library(biomaRt)
mmul = useMart(biomart = "ENSEMBL_MART_ENSEMBL",dataset="mmulatta_gene_ensembl")
cayo_genes <- cayo_data$ensembl_gene_id
cayo_gene_names <- getBM(attributes = c('external_gene_name','ensembl_gene_id'), filters = 'ensembl_gene_id', mart = mmul, values = cayo_genes)
cayo_gene_names <- cayo_gene_names[order(cayo_gene_names$external_gene_name),]
cayo_gene_names <- cayo_gene_names[531:nrow(cayo_gene_names),]
cayo_data <- subset(cayo_data, ensembl_gene_id %in% cayo_gene_names$ensembl_gene_id)
cayo_gene_names <- cayo_gene_names[order(cayo_gene_names$ensembl_gene_id),]
cayo_data$gene_symbol <- cayo_gene_names$external_gene_name
cayo_sig_genes <- subset(cayo_data, fdr.age < 0.3)$gene_symbol
```

Results plots of age effects models on ascertainment gene set

Total PBMCs
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/totalPBMC_Means.dashboard_plusPCs.png)
```{r}
load('/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/pseudobulk.age.modelRes.FDR.RData')
load('/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/pseudobulk.age.modelRes.FDR_AGS.RData')
qqplot(-log10(totalPBMC_modelRes_fdr$p_value_age),-log10(totalPBMC_modelRes_fdr_AGS$p_value_age), xlab='-log10(p) all genes', ylab='-log10(p) ascertainment set', main='total PBMCs')
abline(0,1)
```


CD4 T cells
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/CD4_Means.dashboard_AGS.png)
```{r}
qqplot(-log10(CD4_modelRes_fdr$p_value_age),-log10(CD4_modelRes_fdr_AGS$p_value_age), xlab='-log10(p) all genes', ylab='-log10(p) ascertainment set', main='CD4')
abline(0,1)
```

CD8_A T cells
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/CD8_A_Means.dashboard_AGS.png)
```{r}
qqplot(-log10(CD8_A_modelRes_fdr$p_value_age),-log10(CD8_A_modelRes_fdr_AGS$p_value_age), xlab='-log10(p) all genes', ylab='-log10(p) ascertainment set', main='CD8_A')
abline(0,1)
```

CD8_B T cells
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/CD8_B_Means.dashboard_AGS.png)
```{r}
qqplot(-log10(CD8_B_modelRes_fdr$p_value_age),-log10(CD8_B_modelRes_fdr_AGS$p_value_age), xlab='-log10(p) all genes', ylab='-log10(p) ascertainment set', main='CD8_B')
abline(0,1)
```

NK cells
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/NK_Means.dashboard_AGS.png)
```{r}
qqplot(-log10(NK_modelRes_fdr$p_value_age),-log10(NK_modelRes_fdr_AGS$p_value_age), xlab='-log10(p) all genes', ylab='-log10(p) ascertainment set', main='NK')
abline(0,1)
```

B cells
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/Bcell_Means.dashboard_AGS.png)
```{r}
qqplot(-log10(Bcells_modelRes_fdr$p_value_age),-log10(Bcells_modelRes_fdr_AGS$p_value_age), xlab='-log10(p) all genes', ylab='-log10(p) ascertainment set', main='B cells')
abline(0,1)
```

CD14high monocytes
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/CD14high_Means.dashboard_AGS.png)
```{r}
qqplot(-log10(CD14high_modelRes_fdr$p_value_age),-log10(CD14high_modelRes_fdr_AGS$p_value_age), xlab='-log10(p) all genes', ylab='-log10(p) ascertainment set', main='CD14high mono')
abline(0,1)
```

CD14low monocytes
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/CD14low_Means.dashboard_AGS.png)
```{r}
qqplot(-log10(CD14low_modelRes_fdr$p_value_age),-log10(CD14low_modelRes_fdr_AGS$p_value_age), xlab='-log10(p) all genes', ylab='-log10(p) ascertainment set', main='CD14low mono')
abline(0,1)
```

To summarize, there is a signal of age effects on the total PBMCs (this model also included PCs 1 & 2 of cell type composition), but not for the individual cell types. 

It also looks as though the ascertainment set is enriched for lower p values across the distribution.
QQplot:
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/totalPBMC_allGenesVascertainmentsetQQplot.png){width=50%}


Another approach taken was to use PC1.agg-associated genes from SGE as the ascertainment set to ask about overlap between aging and PC1.agg. For each cell type in SGE we selected the intersection set of genes between NC and LPS with an p < 0.01 (with the exception of CD8_B, where I used a less restrictive nominal p value of < 0.05 to get more genes). 

To summarize the results of this approach, we do not see a signature of age effects in the ascertainment set (based on PC1.agg effects)

Results plots

CD4
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/CD4_Means.dashboard_AGS_SGE.png)
```{r}
load('/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/pseudobulk.age.modelRes.FDR_AGS_SGE.RData')
qqplot(-log10(CD4_modelRes_fdr$p_value_age),-log10(CD4_modelRes_fdr_AGS_SGE$p_value_age), xlab='-log10(p) all genes', ylab='-log10(p) ascertainment set', main='CD4')
abline(0,1)
```

CD8_A
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/CD8_A_Means.dashboard_AGS_SGE.png)
```{r}
qqplot(-log10(CD8_A_modelRes_fdr$p_value_age),-log10(CD8_A_modelRes_fdr_AGS_SGE$p_value_age), xlab='-log10(p) all genes', ylab='-log10(p) ascertainment set', main='CD8_A')
abline(0,1)
```

CD8_B
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/CD8_B_Means.dashboard_AGS_SGE.png)
```{r}
qqplot(-log10(CD8_B_modelRes_fdr$p_value_age),-log10(CD8_B_modelRes_fdr_AGS_SGE$p_value_age), xlab='-log10(p) all genes', ylab='-log10(p) ascertainment set', main='CD8_A')
abline(0,1)
```

NK
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/NK_Means.dashboard_AGS_SGE.png)
```{r}
qqplot(-log10(NK_modelRes_fdr$p_value_age),-log10(NK_modelRes_fdr_AGS_SGE$p_value_age), xlab='-log10(p) all genes', ylab='-log10(p) ascertainment set', main='CD8_A')
abline(0,1)
```

B cell
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/Bcell_Means.dashboard_AGS_SGE.png)
```{r}
qqplot(-log10(Bcells_modelRes_fdr$p_value_age),-log10(Bcells_modelRes_fdr_AGS_SGE$p_value_age), xlab='-log10(p) all genes', ylab='-log10(p) ascertainment set', main='CD8_A')
abline(0,1)
```

CD14high mono
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/CD14high_Means.dashboard_AGS_SGE.png)
```{r}
qqplot(-log10(CD14high_modelRes_fdr$p_value_age),-log10(CD14high_modelRes_fdr_AGS_SGE$p_value_age), xlab='-log10(p) all genes', ylab='-log10(p) ascertainment set', main='CD8_A')
abline(0,1)
```

CD14low mono
![](/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/CD14low_Means.dashboard_AGS_SGE.png)
```{r}
qqplot(-log10(CD14low_modelRes_fdr$p_value_age),-log10(CD14low_modelRes_fdr_AGS_SGE$p_value_age), xlab='-log10(p) all genes', ylab='-log10(p) ascertainment set', main='CD8_A')
abline(0,1)
```

### MASH analysis
The second approach taken to address the power limitations in detecting age effects is to leverage information from across all of the individual cell types using a meta-analytic approach (MASH) to determine if we see a shared signature of age across cell types. 

The assess the shared effects will use the model results from our full age effects models (not the ascertainment sets). 

```{r}
library(mashr)
library(devtools)
library(ashr)
library(gplots)
library(plyr)
library(tidyverse)
```
```{r}
load('/Users/noah/Desktop/SGE_behavioral_project_git/SGE_Behavioral_Project/pseudobulk.age.modelRes.FDR.RData')
Bhat <- list(CD4_modelRes_fdr,CD8_A_modelRes_fdr,CD8_B_modelRes_fdr,NK_modelRes_fdr,CD14high_modelRes_fdr,CD14low_modelRes_fdr,Bcells_modelRes_fdr) %>% 
  map(~ .x %>% 
        as.data.frame %>%
        rownames_to_column('rn')) %>% 
  reduce(full_join, by = 'rn') %>%
  column_to_rownames('rn')

Bhat <- Bhat[,seq(2,ncol(Bhat),7)]
Bhat <- na.omit(Bhat)
colnames(Bhat) <- c('CD4','CD8_A','CD8_B','NK','CD14high','CD14low','Bcells')

Shat <- list(CD4_modelRes_fdr,CD8_A_modelRes_fdr,CD8_B_modelRes_fdr,NK_modelRes_fdr,CD14high_modelRes_fdr,CD14low_modelRes_fdr,Bcells_modelRes_fdr) %>% 
  map(~ .x %>% 
        as.data.frame %>%
        rownames_to_column('rn')) %>% 
  reduce(full_join, by = 'rn') %>%
  column_to_rownames('rn')

Shat <- Shat[,seq(4,ncol(Shat),7)]
Shat <- na.omit(Shat)
colnames(Shat) <- c('CD4','CD8','NK','CD14high','CD14low','Bcells')
Shat <- sqrt(Shat)


#Distribution of Betas across species/time
celltype<-c('CD4','CD8_A','CD8_B','NK','CD14high','CD14low','Bcells')
```

Plot distribution of betas and varbetas across cell types 
```{r}
ggplot()+
  geom_density(aes(Bhat[,1],col=celltype[1]))+
  geom_density(aes(Bhat[,2],col=celltype[2]))+
  geom_density(aes(Bhat[,3],col=celltype[3]))+
  geom_density(aes(Bhat[,4],col=celltype[4]))+
  geom_density(aes(Bhat[,5],col=celltype[5]))+
  geom_density(aes(Bhat[,6],col=celltype[6]))+
  geom_density(aes(Bhat[,7],col=celltype[7]))+
  xlab("betas")


#Distribution of standard error of Betas across species/time

ggplot()+
  geom_density(aes(Shat[,1],col=celltype[1]))+
  geom_density(aes(Shat[,2],col=celltype[2]))+
  geom_density(aes(Shat[,3],col=celltype[3]))+
  geom_density(aes(Shat[,4],col=celltype[4]))+
  geom_density(aes(Shat[,5],col=celltype[5]))+
  geom_density(aes(Shat[,6],col=celltype[6]))+
  geom_density(aes(Shat[,7],col=celltype[7]))+
  xlab("var(betas)")
```

Run mashr. Here I am using both canonical and data-driven covariance matrices, as suggested by mashr vignette.
```{r, message=FALSE, warning=FALSE}
# run mashr
data = mash_set_data(as.matrix(Bhat),as.matrix(Shat))
m.1by1 = mash_1by1(data)
strong = get_significant_results(m.1by1,0.1)

U.pca = cov_pca(data,5,subset=strong)
#print(names(U.pca))

U.ed = cov_ed(data, U.pca, subset=strong)

U.c = cov_canonical(data)
m = mash(data, c(U.c,U.ed))
#print(get_loglik(m),digits = 10)
```
Let's first look at the mixture proportions of the covariance matrices:
```{r}
barplot(get_estimated_pi(m)[3:9],las = 2)
```

Now we can ask how many significant genes have significant age effects in a least one cell types, and by extension how many genes are shared across all cell types.

Number of effects significant in at least one condition, across multiple significance thresholds:
```{r}
m.sig <- get_significant_results(m, thresh=0.05, sig_fn=get_lfsr); length(m.sig)
m.sig <- get_significant_results(m, thresh=0.1, sig_fn=get_lfsr); length(m.sig)
m.sig <- get_significant_results(m, thresh=0.2, sig_fn=get_lfsr); length(m.sig)
```

We can now count the number of conditions in which each gene shows significant age effects:
```{r}
nsig = get_n_significant_conditions(m, thresh = 0.1)
barplot(table(nsig)[-1], ylim = c(0,350), xlab = '# of cell types', ylab = '# of genes')
```

As expected we see decreased sharing as we increase the number of cell types. At the most extreme end of sharing we have 29 genes that have significant age effects in all seven cell types. 

Next we can ask the degree to which sharing occurs across cell type with respect to sign and magnitude of age effects. First, looking at the proportion of all genes that share sign across cell types:
```{r, message=FALSE}
library(corrplot)
x = get_pairwise_sharing(m, factor=0, lfsr_thresh = 10)
corrplot(x, method='color', cl.lim=c(0,1), type='upper', addCoef.col = "black", tl.col="black", tl.srt=45, title = 'Pairwise Sharing by sign', mar = c(4,0,4,0))
```

We can also look at sign sharing of only significant genes:
```{r, message=FALSE}
x = get_pairwise_sharing(m, factor=0, lfsr_thresh = 0.05)
corrplot(x, method='color', cl.lim=c(0,1), type='upper', addCoef.col = "black", tl.col="black", tl.srt=45, title = 'Pairwise Sharing by sign', mar = c(4,0,4,0))
```

As expected the proportions of sign sharing increases when looking only at significant genes. We see a high level of sign sharing across cell types, particularly within the T/NK cells.

Now we can include magnitude, so here sharing is defined as the same sign and similar in magnitude, where similar in magnitude is defined as within a factor of 0.5 of one another. 
```{r}
x = get_pairwise_sharing(m, factor=0.5, lfsr_thresh = 0.1)
corrplot(x, method='color', cl.lim=c(0,1), type='upper', addCoef.col = "black", tl.col="black", tl.srt=45, title = 'Pairwise sharing by sign and magnitude', mar = c(4,0,4,0))
```

One results that stands out is that the smaller CD8 cluster (CD8_B) does not share effects as strongly as you would think given the level of sharing between the other T cell cluster and NK cells. 

For the shared effects, let's ask whether genes with age effects in at least one cell type overlap with age-effect genes from the Cayo ascertainment set or the SGE PC1.agg ascertainment set.
First, cayo set:
```{r}
paste('nGenes overlapping: ', sum(names(m.sig) %in% cayo_sig_genes))
fisher.test(matrix(c(7548, 438, 479, 12), nrow=2), alternative="greater")
```

The number of genes overlapping between MASH and the cayo ascertainment set is not significant. 

Now looking at overlap with the SGE PC1.agg significant genes
```{r}
CD4.df <- readRDS('/Users/noah/Downloads/res_fdr_celltype_CD4T_cells_covariate_PC1.agg_nperms_100.RDS')
CD8_A.df <- readRDS('/Users/noah/Downloads/res_fdr_celltype_0_covariate_PC1.agg_nperms_100.RDS')
CD8_B.df <- readRDS('/Users/noah/Downloads/res_fdr_celltype_3_covariate_PC1.agg_nperms_100.RDS')
CD14hi.df <- readRDS('/Users/noah/Downloads/res_fdr_celltype_mono.CD14hi_covariate_PC1.agg_nperms_100.RDS')
CD14low.df <- readRDS('/Users/noah/Downloads/res_fdr_celltype_mono.CD14low_covariate_PC1.agg_nperms_100.RDS')
NK.df <- readRDS('/Users/noah/Downloads/res_fdr_celltype_NK_covariate_PC1.agg_nperms_100.RDS')
B.df <- readRDS('/Users/noah/Downloads/res_fdr_celltype_B_cells_covariate_PC1.agg_nperms_100.RDS')
```

CD4
```{r}
paste('nGenes overlapping: ', sum(names(m.sig) %in% unique(c(rownames(subset(CD4.df, `p_value_trtNC:PC1.agg` < 0.01)), rownames(subset(CD4.df, `p_value_trtLPS:PC1.agg` < 0.01))))))
totalGeneNum <- nrow(CD4.df)
sge_sig <- length(unique(c(rownames(subset(CD4.df, `p_value_trtNC:PC1.agg` < 0.01)), rownames(subset(CD4.df, `p_value_trtLPS:PC1.agg` < 0.01)))))
mash_sig <- length(m.sig)
overlap <- sum(names(m.sig) %in% unique(c(rownames(subset(CD4.df, `p_value_trtNC:PC1.agg` < 0.01)), rownames(subset(CD4.df, `p_value_trtLPS:PC1.agg` < 0.01)))))
fisher.test(matrix(c(totalGeneNum - c(sge_sig+mash_sig+overlap), sge_sig, mash_sig, overlap), nrow=2), alternative="greater")
```

CD8_A
```{r}
paste('nGenes overlapping: ', sum(names(m.sig) %in% unique(c(rownames(subset(CD8_A.df, `qvalue_trtNC:PC1.agg` < 0.1)), rownames(subset(CD8_A.df, `qvalue_trtLPS:PC1.agg` < 0.1))))))
totalGeneNum <- nrow(CD8_A.df)
sge_sig <- length(unique(c(rownames(subset(CD8_A.df, `qvalue_trtNC:PC1.agg` < 0.1)), rownames(subset(CD8_A.df, `qvalue_trtLPS:PC1.agg` < 0.1)))))
mash_sig <- length(m.sig)
overlap <- sum(names(m.sig) %in% unique(c(rownames(subset(CD8_A.df, `qvalue_trtNC:PC1.agg` < 0.1)), rownames(subset(CD8_A.df, `qvalue_trtLPS:PC1.agg` < 0.1)))))
fisher.test(matrix(c(totalGeneNum - c(sge_sig+mash_sig+overlap), sge_sig, mash_sig, overlap), nrow=2), alternative="greater")
```

CD8_B
```{r}
paste('nGenes overlapping: ', sum(names(m.sig) %in% unique(c(rownames(subset(CD8_B.df, `p_value_trtNC:PC1.agg` < 0.01)), rownames(subset(CD8_B.df, `p_value_trtLPS:PC1.agg` < 0.01))))))
totalGeneNum <- nrow(CD8_B.df)
sge_sig <- length(unique(c(rownames(subset(CD8_B.df, `p_value_trtNC:PC1.agg` < 0.01)), rownames(subset(CD8_B.df, `p_value_trtLPS:PC1.agg` < 0.01)))))
mash_sig <- length(m.sig)
overlap <- sum(names(m.sig) %in% unique(c(rownames(subset(CD8_B.df, `p_value_trtNC:PC1.agg` < 0.01)), rownames(subset(CD8_B.df, `p_value_trtLPS:PC1.agg` < 0.01)))))
fisher.test(matrix(c(totalGeneNum - c(sge_sig+mash_sig+overlap), sge_sig, mash_sig, overlap), nrow=2), alternative="greater")
```

NK
```{r}
paste('nGenes overlapping: ', sum(names(m.sig) %in% unique(c(rownames(subset(NK.df, `qvalue_trtNC:PC1.agg` < 0.1)), rownames(subset(NK.df, `qvalue_trtLPS:PC1.agg` < 0.1))))))
totalGeneNum <- nrow(NK.df)
sge_sig <- length(unique(c(rownames(subset(NK.df, `qvalue_trtNC:PC1.agg` < 0.1)), rownames(subset(NK.df, `qvalue_trtLPS:PC1.agg` < 0.1)))))
mash_sig <- length(m.sig)
overlap <- sum(names(m.sig) %in% unique(c(rownames(subset(NK.df, `qvalue_trtNC:PC1.agg` < 0.1)), rownames(subset(NK.df, `qvalue_trtLPS:PC1.agg` < 0.1)))))
fisher.test(matrix(c(totalGeneNum - c(sge_sig+mash_sig+overlap), sge_sig, mash_sig, overlap), nrow=2), alternative="greater")
```

B
```{r}
paste('nGenes overlapping: ', sum(names(m.sig) %in% unique(c(rownames(subset(B.df, `p_value_trtNC:PC1.agg` < 0.01)), rownames(subset(B.df, `p_value_trtLPS:PC1.agg` < 0.01))))))
totalGeneNum <- nrow(B.df)
sge_sig <- length(unique(c(rownames(subset(B.df, `p_value_trtNC:PC1.agg` < 0.01)), rownames(subset(B.df, `p_value_trtLPS:PC1.agg` < 0.01)))))
mash_sig <- length(m.sig)
overlap <- sum(names(m.sig) %in% unique(c(rownames(subset(B.df, `p_value_trtNC:PC1.agg` < 0.01)), rownames(subset(B.df, `p_value_trtLPS:PC1.agg` < 0.01)))))
fisher.test(matrix(c(totalGeneNum - c(sge_sig+mash_sig+overlap), sge_sig, mash_sig, overlap), nrow=2), alternative="greater")
```

CD14hi
```{r}
paste('nGenes overlapping: ', sum(names(m.sig) %in% unique(c(rownames(subset(CD14hi.df, `qvalue_trtNC:PC1.agg` < 0.1)), rownames(subset(CD14hi.df, `qvalue_trtLPS:PC1.agg` < 0.1))))))
totalGeneNum <- nrow(CD14hi.df)
sge_sig <- length(unique(c(rownames(subset(CD14hi.df, `qvalue_trtNC:PC1.agg` < 0.1)), rownames(subset(CD14hi.df, `qvalue_trtLPS:PC1.agg` < 0.1)))))
mash_sig <- length(m.sig)
overlap <- sum(names(m.sig) %in% unique(c(rownames(subset(CD14hi.df, `qvalue_trtNC:PC1.agg` < 0.1)), rownames(subset(CD14hi.df, `qvalue_trtLPS:PC1.agg` < 0.1)))))
fisher.test(matrix(c(totalGeneNum - c(sge_sig+mash_sig+overlap), sge_sig, mash_sig, overlap), nrow=2), alternative="greater")
```

CD14low
```{r}
paste('nGenes overlapping: ', sum(names(m.sig) %in% unique(c(rownames(subset(CD14low.df, `qvalue_trtNC:PC1.agg` < 0.1)), rownames(subset(CD14low.df, `qvalue_trtLPS:PC1.agg` < 0.1))))))
totalGeneNum <- nrow(CD14low.df)
sge_sig <- length(unique(c(rownames(subset(CD14low.df, `qvalue_trtNC:PC1.agg` < 0.1)), rownames(subset(CD14low.df, `qvalue_trtLPS:PC1.agg` < 0.1)))))
mash_sig <- length(m.sig)
overlap <- sum(names(m.sig) %in% unique(c(rownames(subset(CD14low.df, `qvalue_trtNC:PC1.agg` < 0.1)), rownames(subset(CD14low.df, `qvalue_trtLPS:PC1.agg` < 0.1)))))
fisher.test(matrix(c(totalGeneNum - c(sge_sig+mash_sig+overlap), sge_sig, mash_sig, overlap), nrow=2), alternative="greater")
```
